mpi-proxy-split/record-replay.h

```diff
@@ -106,34 +107,22 @@ namespace dmtcp_mpi
   // Struct for saving arbitrary function arguments
   struct FncArg
   {
-    void *_data;
+    std::shared_ptr<void> _data;
     enum TYPE _type;
 
-    FncArg(const void *data, size_t len, dmtcp_mpi::TYPE type)
-      : _data(JALLOC_HELPER_MALLOC(len))
+    // Default _type set to TYPE_INT_ARRAY because this constructor is also used
+    // by CREATE_LOG_BUF in MPI_Cart functions.
+    FncArg(const void *data, size_t len, dmtcp_mpi::TYPE type = TYPE_INT_ARRAY)
+      : _type(type)
     {
-      _type = type;
-      if (_data && data) {
-        memcpy(_data, data, len);
+      void *ptr = JALLOC_HELPER_MALLOC(len);
+      if (ptr) {
+        _data = std::shared_ptr<void>(ptr,
+                                      [](void *p) { JALLOC_HELPER_FREE(p);});
       }
-    }
 
-    // This constructor is only used by CREATE_LOG_BUF
-    FncArg(const void *data, size_t len)
-      : _data(JALLOC_HELPER_MALLOC(len))
-    {
-      // Default _type set to TYPE_INT_ARRAY because this constructor is used
-      // by CREATE_LOG_BUF in MPI_Cart functions.
-      _type = dmtcp_mpi::TYPE_INT_ARRAY;
       if (_data && data) {
-        memcpy(_data, data, len);
-      }
-    }
-
-    ~FncArg()
-    {
-      if (!_data) {
```

Comment:
This doesn't matter now that we are revising this code.  But isn't `if (!_data)` a separate original bug (a memory leak)?

First of all, `_data` will always be a valid pointer.  All versions of the constructor allocated a _data buffer.  Maybe this logic is left over from an earlier design.  So, it seems like `!_data` will always be false, and so we never free `_data`.  This is a memory leak.